### DISCUSSION TITLE
AI Assist to create Typespec configurations from natural language inside VSCODE.

### DISCUSSION THREAD

**Discussion by samyakkkk** [microsoft/typespec#3348] [Ideas]
Hi everyone! 

I've been working on programming agents recently and came across Typespec. 

Been thinking if it would be interesting to have an AI agent inside VSCode that can assist in creating various kinds of TypeSpec configurations?

If anybody is interested, we can try and create something :) (also happy to collaborate if anybody is interested to contribute). 

Look forward!

**Comment by markcowl** [comment]
@mario-guerra thoughts?

**Comment by mario-guerra** [comment]
Hi @samyakkkk, this sounds interesting. I'd like to know more, can you elaborate on what you have in mind?

**Reply by samyakkkk** [reply]
hi @mario-guerra! just a simple AI agent that is trained on the latest docs and examples of Typespec. Devs would be able to use it in VSCode via a chat interface and ask to generate configurations from natural language.

I recently launched a tool for this - https://www.commanddash.io. If the team is fine with it, we can add a Typespec agent and share it here for the community to try! 

Plmk :)

**Comment by mario-guerra** [comment]
Hi @samyakkkk, thanks for the link. I checked out the project and it looks interesting. I'd like to know more about what you mean by an agent that is trained on the latest docs, are you actually training a model or are you using a retrieval-augmented generation pattern to supplement the responses produced by Gemini? 

Either way I think this would be interesting to pursue, what do you need from us to move forward?

**Comment by samyakkkk** [comment]
@mario-guerra great! we will be using RAG which I think should suffice the needs for now. 

Will you be able to help with some datasources we should consider indexing beyond the docs: https://typespec.io/docs?

Like examples tsp configuration files - to use as few-shot examples for model. If not available, we can try to generate synthetically. 

thanks in advance.

**Comment by mario-guerra** [comment]
Thanks for clarifying @samyakkkk. The docs are a great start, as for extra data sources I suppose that depends on what you mean by TypeSpec configuration files, can you elaborate?

Azure services are in the process of converting to TypeSpec, so we have TypeSpec specs in our [specs repo](https://github.com/Azure/azure-rest-api-specs/tree/main/specification), but they are specific to Azure. I'm not aware of a 3P TSP specs repo I can point you to.

**Comment by samyakkkk** [comment]
Hi @mario-guerra! Please find the Typespec agent indexed on the docs 🎉 

<img width="380" alt="Screenshot 2024-06-28 at 9 57 11 PM" src="https://github.com/microsoft/typespec/assets/56000318/c0c5e585-8806-4298-8903-862ddc330f58">

I asked it  to **_"Write me a User API definition with create and update endpoints. Also accept profile picture of the user"_** and got the below response:

<img width="376" alt="Screenshot 2024-06-28 at 9 49 43 PM" src="https://github.com/microsoft/typespec/assets/56000318/db4696b8-b040-443b-bf92-4f1f43bd74ae">

Complete code:
```
import "@typespec/http";

using TypeSpec.Http;

@service({
  title: "User Service",
  version: "1.0.0"
})
@server("https://api.example.com", "Production server")
namespace UserService {

  // Define the User model
  model User {
    id: string;
    username: string;
    email: string;
    profilePicture: bytes;
  }

  // Define the create user operation
  @route("/users")
  @post
  op createUser(
    @header contentType: "multipart/form-data",
    @body username: string,
    @body email: string,
    @body profilePicture: bytes
  ): {
    @statusCode statusCode: 201;
    @body user: User;
  };

  // Define the update user operation
  @route("/users/{userId}")
  @put
  op updateUser(
    @path userId: string,
    @header contentType: "multipart/form-data",
    @body username?: string,
    @body email?: string,
    @body profilePicture?: bytes
  ): {
    @statusCode statusCode: 200;
    @body user: User;
  };
}
```

Please test it by installing the agent from the [VSCode extension](https://marketplace.visualstudio.com/items?itemName=WelltestedAI.fluttergpt). Based on what you think - we can improve it further or share with the community!

**Comment by mario-guerra** [comment]
Hi @samyakkkk, I copied the complete code and pasted it into the typespec.io developer playground, the code won't compile due to errors. You can check it for yourself using [this link](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7DQoNCnVzaW5nIFR5cGVTcGVjLkh0dHDFGEBzZXJ2aWNlKHsNCiAgdGl0bGU6ICJVc2VyIFPGGiIsxBp2ZXJzaW9uOiAiMS4wLjAiDQp9Kcc%2BZXIoIsRpczovL2FwaS5leGFtcGxlLmNvbSIsICJQcm9kdWN0aW9uIMYuIikNCm5hbWVzcGFjZSDEa8dqIHvkAJIgIC8vIERlZmluZSB0aMYhIG1vZGVsxBzFCcYU5QCyICBpZDogc3RyaW5nO8YRdXNlcsRgzxdlbWFpbM8UcHJvZmlsZVBpY3R1cmU6IGJ5dGVzxRx99ACLY3JlYXRlxWEgb3BlcmHkANbEJ0Byb3V0ZSgiL8Qbc%2BQA4SAgQHBvc3TECW9wxznkALYoxnlAaGVhZGVyIGNvbnRlbnTkAZg6ICJtdWx0aXBhcnQvZm9ybS1kYXRh5gF%2BICBAYm9kefEA280d7QDhzRr1AOfEISk66AFGQHN0YXR1c0NvZGUgygs6IDIwMecBLOoAgjrlAYDFF33lAkXwASh1cGT%2FASjEG3Mve8QHSWR96AExdegBMMZB7AEwcGF0aMVRSekB7%2BgA4%2F8BS%2F8BSz%2F6AUzWG%2B4BTT%2F%2FAU7wAU4w%2FwFOCn0%3D&e=%40typespec%2Fopenapi3&options=%7B%7D).

**Comment by samyakkkk** [comment]
Thanks for pointing this out @mario-guerra. Tried sending the errors to the agent and got this updated code which worked :)


```
import "@typespec/http";
using TypeSpec.Http;

@service({
  title: "User Service",
})
@server("https://api.example.com", "Production server")
namespace UserService {
  // Define the User model
  model User {
    id: string;
    username: string;
    email: string;
    profilePicture: bytes;
  }

  // Define the CreateUserRequest model
  model CreateUserRequest {
    username: string;
    email: string;
    profilePicture: bytes;
  }

  // Define the UpdateUserRequest model
  model UpdateUserRequest {
    username?: string;
    email?: string;
    profilePicture?: bytes;
  }

  // Define the create user operation
  @route("/users")
  @post
  op createUser(
    @header contentType: "multipart/form-data",
    @body request: CreateUserRequest
  ): {
    @statusCode statusCode: 201;
    @body user: User;
  };

  // Define the update user operation
  @route("/users/{userId}")
  @put
  op updateUser(
    @path userId: string,
    @header contentType: "multipart/form-data",
    @body request: UpdateUserRequest
  ): {
    @statusCode statusCode: 200;
    @body user: User;
  };
}
```

RAG alone on docs is a little weak in generating working code in one-shot. For this reason, I was requesting if we have any source of various example TSP files covering variety use cases. It would really bump up the zero shot accuracy.

**Reply by samyakkkk** [reply]
But even without the examples - the agent can be really helpful in asking direct questions on the docs from the IDE and getting contextualized help - please give it a shot if you get a chance.

**Comment by mario-guerra** [comment]
I will make some time next week to try it out, thanks for putting this together!

**Comment by samyakkkk** [comment]
@mario-guerra we now also have the Typespec agent available as in web https://app.commanddash.io/agent/typespec (no setup). 

You can now easily give it a try!

**Comment by samyakkkk** [comment]
@mario-guerra just circling back on if you could test the codegen agent :) Been looking forward to your feedback! 

if you find it helpful, it could be great to highlight this for the community.

**Reply by samyakkkk** [reply]
It's available without any setup here: https://app.commanddash.io/agent/typespec
